_posts/2019-07-03-optimize-write-to-neo4-recommendation-engine.md

```diff
@@ -0,0 +1,196 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Consumer;
+
+use App\Message\Command\Advert\AdvertServedToUser;
+use App\Message\Command\Advert\RunAdvertServedToUserBatch;
+use OldSound\RabbitMqBundle\RabbitMq\BatchConsumerInterface;
+use PhpAmqpLib\Message\AMQPMessage;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Messenger\MessageBusInterface;
+
+class AdvertServedToUserToUserBatchConsumer implements BatchConsumerInterface
+{
+    private $commandBus;
+    private $logger;
+
+    public function __construct(MessageBusInterface $commandBus, LoggerInterface $logger)
+    {
+        $this->commandBus = $commandBus;
+        $this->logger = $logger;
+    }
+
+    /**
+     * @param AMQPMessage[] $messages
+     *
+     * @return array|bool
+     */
+    public function batchExecute(array $messages)
+    {
+        try {
+            $command = new RunAdvertServedToUserBatch($messages);
+            $this->commandBus->dispatch($command);
+            return true;
+        } catch (\Throwable $e) {
+            $this->logger->log('error', $e->getMessage(), ['exception' => $e]);
+            return false;
+        } 
+    }
+}
+{% endhighlight %}
+
+registered class as a service:
+
+{% highlight yaml %}
+#config/services.yaml
+services:
+    ...
+    app.consumer.batch:
+        class: App\Consumer\AdvertServedToUserBatchConsumer
+        arguments: ["@messenger.bus.command", '@logger']
+{% endhighlight %}
+
+and use our consumer as callback for batch RabbitMQ
+
+{% highlight yaml %}
+#config/packages/old_sound_rabbit_mq.yaml
+old_sound_rabbit_mq:
+    connections:
+        default:
+            ...
+        advert_served_to_user:
+            url: '%env(BATCH_MESSAGES_TRANSPORT_DSN)%'
+            lazy:     true
+            connection_timeout: 3
+            read_write_timeout: 3
+            # requires php-amqplib v2.4.1+ and PHP5.4+
+            keepalive: false
+            # requires php-amqplib v2.4.1+
+            heartbeat: 0
+    consumers:
+        ...
+    	batch_consumers:
+        advert_served_to_user:
+            connection:       advert_served_to_user
+            exchange_options: {name: 'advert-served-to-user', type: fanout }
+            queue_options:    {name: 'advert-served-to-user'}
+            callback:         'app.consumer.batch'
+            qos_options:      {prefetch_size: 0, prefetch_count: 5000, global: false}
+            timeout_wait:     5
+            auto_setup_fabric: false
+            idle_timeout_exit_code: -2
+            keep_alive: false
+            graceful_max_execution:
+                timeout: 60
+{% endhighlight %}
+
+After that we created a handler for our batch command that uses UNWIND query:
+
+{% highlight php %}
+<?php
+
+namespace App\Message\CommandHandler\Advert;
+
+use App\Message\Command\Advert\RunAdvertServedToUserBatch;
+use App\Message\CommandHandler\BaseCommandHandler;
+use App\Model\AdvertServedToUserM
+
+class RunAdvertServedToUserBatchHandler extends BaseCommandHandler
```

Comment:
We could (for the sake of the example) move this code to the consumer. That may make it easier to understand. What do you think=

Reply:
sure